pimd/pim6_mroute.c

```diff
@@ -0,0 +1,1125 @@
+	if (up->rpf.source_nexthop.interface &&
+	    up->channel_oil->oil.mf6cc_parent >= MAXMIFS) {
+		pim_upstream_mroute_iif_update(up->channel_oil, __func__);
+	}
+	pim_register_join(up);
+	/* if we have receiver, inherit from parent */
+	pim_upstream_inherited_olist_decide(pim_ifp->pim, up);
+
+	return 0;
+}
+
+static int pim_mroute_msg_wholepkt(int fd, struct interface *ifp,
+				   const char *buf)
+{
+	struct pim_interface *pim_ifp;
+	pim_sgaddr sg;
+	struct pim_rpf *rpg;
+	const struct ip6_hdr *ip6_hdr;
+	struct pim_upstream *up;
+
+	pim_ifp = ifp->info;
+
+	ip6_hdr = (const struct ip6_hdr *)buf;
+
+	memset(&sg, 0, sizeof(pim_sgaddr));
+	IPV6_ADDR_COPY (&sg.src, &(ip6_hdr->ip6_src));
+	IPV6_ADDR_COPY (&sg.grp, &(ip6_hdr->ip6_dst));
+
+	up = pim_upstream_find(pim_ifp->pim, &sg);
+	if (!up) {
+		pim_sgaddr star = sg;
+		star.src = in6addr_any;
+
+		up = pim_upstream_find(pim_ifp->pim, &star);
+
+		if (up && PIM_UPSTREAM_FLAG_TEST_CAN_BE_LHR(up->flags)) {
+			up = pim_upstream_add(pim_ifp->pim, &sg, ifp,
+					      PIM_UPSTREAM_FLAG_MASK_SRC_LHR,
+					      __func__, NULL);
+			if (!up) {
+				if (PIM_DEBUG_MROUTE)
+					zlog_debug(
+						"%s: Unable to create upstream information for %s",
+						__func__, pim_str_sg_dump(&sg));
+				return 0;
+			}
+			pim_upstream_keep_alive_timer_start(
+				up, pim_ifp->pim->keep_alive_time);
+			pim_upstream_inherited_olist(pim_ifp->pim, up);
+			pim_upstream_update_join_desired(pim_ifp->pim, up);
+
+			if (PIM_DEBUG_MROUTE)
+				zlog_debug("%s: Creating %s upstream on LHR",
+					   __func__, up->sg_str);
+			return 0;
+		}
+		if (PIM_DEBUG_MROUTE_DETAIL) {
+			zlog_debug(
+				"%s: Unable to find upstream channel WHOLEPKT%s",
+				__func__, pim_str_sg_dump(&sg));
+		}
+		return 0;
+	}
+
+	if (!up->rpf.source_nexthop.interface) {
+		if (PIM_DEBUG_PIM_TRACE)
+			zlog_debug("%s: up %s RPF is not present", __func__,
+				   up->sg_str);
+		return 0;
+	}
+
+	pim_ifp = up->rpf.source_nexthop.interface->info;
+
+	rpg = pim_ifp ? RP(pim_ifp->pim, sg.grp) : NULL;
+
+	if ((pim_rpf_addr_is_inaddr_none(rpg)) || (!pim_ifp)
+	    || (!(PIM_I_am_DR(pim_ifp)))) {
+		if (PIM_DEBUG_MROUTE) {
+			zlog_debug("%s: Failed Check send packet", __func__);
+		}
+		return 0;
+	}
+
+	/*
+	 * If we've received a register suppress
+	 */
+	if (!up->t_rs_timer) {
+		if (pim_is_grp_ssm(pim_ifp->pim, sg.grp)) {
+			if (PIM_DEBUG_PIM_REG)
+				zlog_debug(
+					"%s register forward skipped as group is SSM",
+					pim_str_sg_dump(&sg));
+			return 0;
+		}
+
+		if (!PIM_UPSTREAM_FLAG_TEST_FHR(up->flags)) {
+			if (PIM_DEBUG_PIM_REG)
+				zlog_debug(
+					"%s register forward skipped, not FHR",
+					up->sg_str);
+			return 0;
+		}
+
+		pim_register_send((uint8_t *)buf + sizeof(struct ip6_hdr),
+				  ntohs(ip6_hdr->ip6_plen) - IPV6_HEADER_LENGTH,
+				  pim_ifp->primary_address, rpg, 0, up);
+	}
+	return 0;
+}
+
+static int pim_mroute_msg_wrongmif(int fd, struct interface *ifp,
```

Comment:
It's better to re-use the same functions if there is little change.

Reply:
The datastructures mfcctl and mf6cctl are different. so reusing the same function for v4 and v6 would be very unlikely